include/zephyr/drivers/led/lp50xx.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ * Copyright (c) 2022 Grinn
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```

Comment:
OK this whole file is not working. I apologize for leading you in this direction.

It is not good to share the chip enumeration between the driver and the user application. The `lp50xx_get_chip_id` function exposed by the driver is a bad idea too. And all this macros depending on the chip type and computed at runtime are not good as well.

I don't think it is correct either to move the chipset abstraction into the driver channel function (idea suggested by @bbilas). This would defeat the purpose of the channel API which is to give raw access to the registers to a user application.

I have been wrapping my head around this problem for some time and I can't find an elegant solution. I think we could try to go with separate channel register definitions for each device type. Would you try that and see how it goes ?

Reply:
FYI: We don't have further effort to develop another functionality and options for this driver that we don't know if it will be sufficient to be merged so for now we have only two solutions, either fix the conflicts, merge it as it is (since we know that is working and it was tested on the real hardware) and improve that in the future or close this PR.

Reply:
@simonguinot any decisions or updates? :)

Reply:
Hi @m-janus and @m-janus,

@Mani-Sadhasivam is the decision maker here. I am only helping with the review.

Here is where I stand:

- In its previous version (i.e. without the `lp50xx_get_chip_id` function exposed and the chip enumeration shared between the driver and the user application), the driver was good.
- But there was still support for the LP5009/12 chips to be added to the sample. And the path we tried with the last attempt was not good. It's unfortunate but setbacks happen sometimes.

That's it. So IMHO you have still some work to do guys.

Now, that's being said, I'd really like to see this PR merged. So if nothing happens, I'll probably take it over.

Simon

Reply:
I don't like this approach of sharing the custom functions and runtime definitions with samples. This all could be simplified with separate samples for each LED chip making use of static definitions.

Reply:
Hi, @simonguinot and @Mani-Sadhasivam

So to summarize and clarify all discussed changes:

1. In your opinion, driver itself is in a good condition now and it doesn't need any changes (except removing `lp50xx_get_chip_id` function)?
2. Only requested changes left are in samples area. In that case sample for LP503X is almost unchanged (only aligned to a new driver) and a new sample for LP5009/12 is created. Are there any specific demands for LP5009/12 sample on your side? Or it could be just almost just a copy-paste sample with similar "show" scenario. I'm wondering if it makes a big sense to create another sample for LP5009/12 as code would be almost same and probably won't bring very useful info for users.

TLDR: Two questions: 1. Driver itself is ready to merge? 2. Does it make sense in your opinion to create another sample for 5009/12.

Thanks in advance ;) After your answers I can decide about following activities.

Reply:
Hi @m-janus,

Yes the driver is almost OK. You need to revert the changes you made for the last update (addition of the `lp50xx_get_chip_id` function, `enum lp50xx_chips` moved into `include/zephyr/drivers/led/lp50xx.h`, etc) and to address the comments made by @Mani-Sadhasivam.

About the sample, I think we can keep all the devices under the same sample. But each device must have static channel register definitions in `include/zephyr/drivers/led/lp50xx.h` (or in separated headers if you prefer). And in the sample, we can have a `run_channel_test` function for the LP5009/12 devices and another for the LP5030/36. The `run_led_test` function should be OK for all the device types. This should work.

Reply:
Thanks for reply. Today I made some changes and I think we're almost there, but I wanted to clarify one, probably last thing. Now reading your comments @simonguinot and @Mani-Sadhasivam I see that you Mani want two separate samples for chips and you Simon want just `run_channel_test` to differ?  I'm asking because putting this into one sample would probably require a lot of #ifdef's or something similar - this is probably not very elegant.
I'm wondering if after latest proposed changes we need to make a sample for LP5009/12? At least it's still one driver ;) Maybe aligning sample for LP5030 to latest changes would be enough? In my opinion this is the best option for now and also big effort was already put on this driver. 
I'd like you both @Mani-Sadhasivam  and @simonguinot to decide for one variant before I change something, to avoid another, unnecessary work.

Also another topic, probably more to you @simonguinot.
In lp50xx.h which option is better in your opinion?
```
/* LED brightness channels. */
#define LP5009_LED_BRIGHT_CHAN_BASE	3
#define LP5009_LED_BRIGHT_CHAN(led)	\
	(LP5009_LED_BRIGHT_CHAN_BASE + led)

#define LP5012_LED_BRIGHT_CHAN_BASE	3
#define LP5012_LED_BRIGHT_CHAN(led)	\
	(LP5012_LED_BRIGHT_CHAN_BASE + led)

```
or something like
```
/* LED brightness channels. */
#define LP5009_12_LED_BRIGHT_CHAN_BASE	3
#define LP5009_12_LED_BRIGHT_CHAN(led)	\
	(LP5009_LED_BRIGHT_CHAN_BASE + led)

```
The first one is more clear I think, but it's some way of duplication. In the second I think the part 09_12 may be misleading. in case of this chip is not as easy as LP5030/LP5036 where we can just put LP503X in name.

Please answer my questions @Mani-Sadhasivam and @simonguinot, so I can put last changes.

Reply:
I'd go for separate samples if stuffing everything in a single sample makes it look messy. Also, IMO having separate samples for each chip is not a big deal compared to drivers.

Reply:
I'd try to keep everything in a single sample, maybe using some `$device.c` files for specific functions. But if the result is not readable enough, I'd fall back to two separate samples. Here we are talking about two test channel functions and some code to identify chipset from DT and switch to the correct test channel function. I could be wrong, but it seems manageable.

About the LP50{09/12} definitions in `lp50xx.h`, I think I would go with the `LP5009_12_` prefix.

Reply:
Hello @simonguinot, hello @Mani-Sadhasivam 

Sorry for such a long break. According to @Mani-Sadhasivam messages, I'll go to two separate samples. Just to verify @Mani-Sadhasivam, do you want to create two another directories in the _samples/drivers/_, e.g. _samples/drivers/led_lp503x_ and _samples/drivers/led_lp5009_?

Reply:
@m-janus Yes. If the end result of clubbing the samples of two devices looks messy (I think it is), I'd prefer to have cleaner two separate samples instead.

Reply:
Hello,
I just pushed updated code. At least I'd like to try one more time with single sample. I hope current proposal is close to the solution @simonguinot mentioned. Let me know if now it's more readable for you. If not, I'll finally go to separate samples.